File: cli-dotnet/ContractLoader.cs
using System.Text.Json;
using Elastic.Clients.Elasticsearch;
using ImportFlights;

namespace ImportContracts;

public class ContractLoader
{
    public const string EsIndex = "contracts";
    public const string PipelineName = "pdf_pipeline";
    public const string DefaultInferenceEndpoint = ".elser-2-elastic";

    private readonly ElasticsearchClientWrapper _client;
    private readonly Dictionary<string, object> _mapping;
    private readonly ILogger _logger;
    private string _inferenceEndpoint;
    private int _indexedCount;

    public ContractLoader(ElasticsearchClientWrapper client, Dictionary<string, object> mapping, ILogger logger, string? inferenceEndpoint = null)
    {
        _client = client;
        _mapping = mapping;
        _logger = logger;
        _inferenceEndpoint = string.IsNullOrWhiteSpace(inferenceEndpoint) ? DefaultInferenceEndpoint : inferenceEndpoint!;
        _indexedCount = 0;
    }

    public async Task<bool> CheckElasticsearchAsync()
    {
        try
        {
            var health = await _client.GetClusterHealthAsync();
            _logger.Info($"Cluster: {health.ClusterName ?? "unknown"}");
            _logger.Info($"Status: {health.Status}");
            return true;
        }
        catch (Exception ex)
        {
            _logger.Error($"Connection error: {ex.Message}");
            return false;
        }
    }

    public async Task<bool> CheckInferenceEndpointAsync()
    {
        try
        {
            var endpoints = await _client.GetInferenceEndpointsAsync();

            // First, try to find the specified endpoint
            var found = endpoints.FirstOrDefault(ep =>
                MatchesEndpoint(ep, _inferenceEndpoint));

            if (found != null)
            {
                _logger.Info($"Found inference endpoint: {_inferenceEndpoint}");
                return true;
            }

            // Auto-detect ELSER endpoints
            var elserEndpoints = endpoints
                .Select(ep => new { Id = EndpointId(ep), Endpoint = ep })
                .Where(ep => ep.Id != null && ep.Id.Contains("elser", StringComparison.OrdinalIgnoreCase))
                .ToList();

            if (elserEndpoints.Count > 0)
            {
                var preferred = elserEndpoints.FirstOrDefault(ep =>
                    ep.Id.Contains(".elser-2-", StringComparison.OrdinalIgnoreCase) ||
                    ep.Id.Contains(".elser_model_2", StringComparison.OrdinalIgnoreCase));

                _inferenceEndpoint = preferred?.Id ?? elserEndpoints.First().Id!;
                _logger.Warn($"Specified endpoint not found, using auto-detected: {_inferenceEndpoint}");
                return true;
            }

            _logger.Error($"Inference endpoint '{_inferenceEndpoint}' not found");
            _logger.Info("Available endpoints:");
            foreach (var ep in endpoints)
            {
                _logger.Info($"  - {EndpointId(ep) ?? "<unknown>"}");
            }
            return false;
        }
        catch (Exception ex)
        {
            _logger.Warn($"Error checking inference endpoint: {ex.Message}");
            _logger.Warn("Continuing anyway...");
            return true;
        }
    }

    public async Task<bool> CreatePipelineAsync()
    {
        var pipelineConfig = new Dictionary<string, object>
        {
            ["description"] = "Extract text from PDF - semantic_text field handles chunking and embeddings",
            ["processors"] = new List<Dictionary<string, object>>
            {
                new()
                {
                    ["attachment"] = new Dictionary<string, object>
                    {
                        ["field"] = "data",
                        ["target_field"] = "attachment",
                        ["remove_binary"] = true
                    }
                },
                new()
                {
                    ["set"] = new Dictionary<string, object>
                    {
                        ["field"] = "semantic_content",
                        ["copy_from"] = "attachment.content",
                        ["ignore_empty_value"] = true
                    }
                },
                new()
                {
                    ["remove"] = new Dictionary<string, object>
                    {
                        ["field"] = "data",
                        ["ignore_missing"] = true
                    }
                },
                new()
                {
                    ["set"] = new Dictionary<string, object>
                    {
                        ["field"] = "upload_date",
                        ["value"] = "{{ _ingest.timestamp }}"
                    }
                }
            }
        };

        try
        {
            await _client.CreatePipelineAsync(PipelineName, pipelineConfig);
            return true;
        }
        catch (Exception ex)
        {
            _logger.Error($"Error creating pipeline: {ex.Message}");
            return false;
        }
    }

    public async Task<bool> CreateIndexAsync()
    {
        try
        {
            if (await _client.IndexExistsAsync(EsIndex))
            {
                _logger.Info($"Deleting existing index '{EsIndex}' before import");
                if (await _client.DeleteIndexAsync(EsIndex))
                {
                    _logger.Info($"Index '{EsIndex}' deleted");
                }
                else
                {
                    _logger.Warn($"Failed to delete index '{EsIndex}'");
                }
            }
        }
        catch (Exception ex)
        {
            _logger.Warn($"Error checking/deleting index: {ex.Message}");
        }

        // Update mapping with detected inference endpoint
        var mappingWithInference = NormalizeToDictionary(_mapping);
        var mappingsObj = EnsureDictionary(mappingWithInference, "mappings");
        var propsObj = EnsureDictionary(mappingsObj, "properties");
        var semanticObj = EnsureDictionary(propsObj, "semantic_content");
        semanticObj["inference_id"] = _inferenceEndpoint;

        _logger.Info($"Creating index: {EsIndex}");
        try
        {
            await _client.CreateIndexAsync(EsIndex, mappingWithInference);
            _logger.Info($"Successfully created index: {EsIndex}");
            return true;
        }
        catch (Exception ex)
        {
            _logger.Error($"Error creating index: {ex.Message}");
            return false;
        }
    }

    public async Task<bool> IngestPdfsAsync(string pdfPath)
    {
        var pdfFiles = GetPdfFiles(pdfPath);
        if (pdfFiles.Count == 0)
        {
            _logger.Error("No PDF files to process");
            return false;
        }

        _logger.Info($"Processing {pdfFiles.Count} PDF file(s)...");

        var success = 0;
        var failed = 0;

        foreach (var pdfFile in pdfFiles)
        {
            if (await IndexPdfAsync(pdfFile))
            {
                success++;
            }
            else
            {
                failed++;
            }
        }

        _logger.Info($"Indexed {success} of {pdfFiles.Count} file(s)");
        if (failed > 0)
        {
            _logger.Warn($"Failed: {failed}");
        }

        return failed == 0;
    }

    public async Task VerifyIngestionAsync()
    {
        await Task.Delay(1000);
        try
        {
            var count = await _client.CountDocumentsAsync(EsIndex);
            _logger.Info($"Index '{EsIndex}' contains {count} document(s)");

            if (_indexedCount > 0 && count == 0)
            {
                _logger.Warn($"Warning: Expected {_indexedCount} document(s) but count shows 0. Documents may have failed during pipeline processing.");
            }
        }
        catch (Exception ex)
        {
            _logger.Warn($"Could not verify document count: {ex.Message}");
        }
    }

    private static string? EndpointId(Dictionary<string, object> endpoint)
    {
        foreach (var key in new[] { "inference_id", "endpoint", "id", "name" })
        {
            if (endpoint.TryGetValue(key, out var value) && value is string s && !string.IsNullOrWhiteSpace(s))
            {
                return s;
            }
        }
        return null;
    }

    private static bool MatchesEndpoint(Dictionary<string, object> endpoint, string target)
    {
        var id = EndpointId(endpoint);
        return id != null && id.Equals(target, StringComparison.Ordinal);
    }

    private static Dictionary<string, object> DeepCopy(Dictionary<string, object> source)
    {
        var json = JsonSerializer.Serialize(source);
        return JsonSerializer.Deserialize<Dictionary<string, object>>(json)!;
    }

    private static Dictionary<string, object> NormalizeToDictionary(Dictionary<string, object> source)
    {
        var normalized = new Dictionary<string, object>();
        foreach (var kvp in source)
        {
            normalized[kvp.Key] = NormalizeValue(kvp.Value);
        }
        return normalized;
    }

    private static Dictionary<string, object>? GetDictionary(Dictionary<string, object>? parent, string key)
    {
        if (parent == null) return null;
        if (!parent.TryGetValue(key, out var value) || value == null) return null;
        return NormalizeValue(value) as Dictionary<string, object>;
    }

    private static Dictionary<string, object> EnsureDictionary(Dictionary<string, object> parent, string key)
    {
        var existing = GetDictionary(parent, key);
        if (existing != null)
        {
            parent[key] = existing;
            return existing;
        }

        var created = new Dictionary<string, object>();
        parent[key] = created;
        return created;
    }

    private static object? NormalizeValue(object? value)
    {
        switch (value)
        {
            case null:
                return null;
            case Dictionary<string, object> dict:
                return NormalizeToDictionary(dict);
            case List<object> list:
                return list.Select(NormalizeValue).ToList();
            case JsonElement element:
                return ConvertJsonElement(element);
            default:
                return value;
        }
    }

    private static object? ConvertJsonElement(JsonElement element)
    {
        switch (element.ValueKind)
        {
            case JsonValueKind.Object:
                var dict = new Dictionary<string, object>();
                foreach (var prop in element.EnumerateObject())
                {
                    dict[prop.Name] = ConvertJsonElement(prop.Value)!;
                }
                return dict;
            case JsonValueKind.Array:
                var list = new List<object>();
                foreach (var item in element.EnumerateArray())
                {
                    var converted = ConvertJsonElement(item);
                    if (converted != null)
                    {
                        list.Add(converted);
                    }
                }
                return list;
            case JsonValueKind.String:
                return element.GetString() ?? string.Empty;
            case JsonValueKind.Number:
                if (element.TryGetInt64(out var l)) return l;
                if (element.TryGetDouble(out var d)) return d;
                return element.GetRawText();
            case JsonValueKind.True:
            case JsonValueKind.False:
                return element.GetBoolean();
            case JsonValueKind.Null:
            case JsonValueKind.Undefined:
            default:
                return null;
        }
    }

    private static string ExtractAirlineName(string filename)
    {
        var lower = filename.ToLowerInvariant();
        if (lower.Contains("american"))
        {
            return "American Airlines";
        }
        if (lower.Contains("southwest"))
        {
            return "Southwest";
        }
        if (lower.Contains("united"))
        {
            return "United";
        }
        if (lower.Contains("delta") || lower.Contains("dl-"))
        {
            return "Delta";
        }
        return "Unknown";
    }

    private List<string> GetPdfFiles(string path)
    {
        if (!File.Exists(path) && !Directory.Exists(path))
        {
            _logger.Error($"Path '{path}' does not exist");
            return new List<string>();
        }

        if (File.Exists(path))
        {
            if (string.Equals(Path.GetExtension(path), ".pdf", StringComparison.OrdinalIgnoreCase))
            {
                return new List<string> { path };
            }

            _logger.Error($"'{path}' is not a PDF file");
            return new List<string>();
        }

        var pdfs = Directory.GetFiles(path, "*.pdf").OrderBy(p => p).ToList();
        if (pdfs.Count == 0)
        {
            _logger.Warn($"No PDF files found in directory '{path}'");
        }
        return pdfs;
    }

    private async Task<bool> IndexPdfAsync(string pdfPath)
    {
        var filename = Path.GetFileName(pdfPath);
        var airline = ExtractAirlineName(filename);

        try
        {
            var pdfBytes = await File.ReadAllBytesAsync(pdfPath);
            var encoded = Convert.ToBase64String(pdfBytes);

            var document = new Dictionary<string, object>
            {
                ["data"] = encoded,
                ["filename"] = filename,
                ["airline"] = airline
            };

            await _client.IndexDocumentAsync(EsIndex, document, PipelineName);

            _logger.Info($"Indexed: {filename} (airline: {airline})");
            _indexedCount += 1;
            return true;
        }
        catch (Exception ex)
        {
            _logger.Error($"Error processing {filename}: {ex.Message}");
            return false;
        }
    }
}

File: cli-java/src/main/java/com/elastic/ContractLoader.java
package com.elastic;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.logging.Logger;
import java.util.stream.Stream;
import java.util.Base64;

public class ContractLoader {
    public static final String ES_INDEX = "contracts";
    public static final String PIPELINE_NAME = "pdf_pipeline";
    public static final String DEFAULT_INFERENCE_ENDPOINT = ".elser-2-elastic";

    private final ElasticsearchClientWrapper client;
    private final Map<String, Object> mapping;
    private final Logger logger;
    private String inferenceEndpoint;
    private int indexedCount;

    public ContractLoader(ElasticsearchClientWrapper client, Map<String, Object> mapping,
                         Logger logger, String inferenceEndpoint) {
        this.client = client;
        this.mapping = mapping;
        this.logger = logger;
        this.inferenceEndpoint = inferenceEndpoint != null ? inferenceEndpoint : DEFAULT_INFERENCE_ENDPOINT;
        this.indexedCount = 0;
    }

    public boolean checkElasticsearch() {
        try {
            co.elastic.clients.elasticsearch.cluster.HealthResponse health = client.clusterHealth();
            String clusterName = health.clusterName() != null ? health.clusterName() : "unknown";
            logger.info("Cluster: " + clusterName);
            logger.info("Status: " + health.status());
            return true;
        } catch (Exception e) {
            logger.severe("Connection error: " + e.getMessage());
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    public boolean checkInferenceEndpoint() {
        try {
            Map<String, Object> response = client.getInferenceEndpoints();
            List<Map<String, Object>> endpoints = (List<Map<String, Object>>) response.get("endpoints");
            if (endpoints == null) {
                endpoints = new ArrayList<>();
            }

            // First, try to find the specified endpoint
            Optional<Map<String, Object>> foundEndpoint = endpoints.stream()
                .filter(ep -> {
                    String id = (String) ep.get("inference_id");
                    return id != null && id.equals(this.inferenceEndpoint);
                })
                .findFirst();

            if (foundEndpoint.isPresent()) {
                logger.info("Found inference endpoint: " + this.inferenceEndpoint);
                return true;
            }

            // Auto-detect ELSER endpoints
            List<Map<String, Object>> elserEndpoints = new ArrayList<>();
            for (Map<String, Object> ep : endpoints) {
                String id = (String) ep.get("inference_id");
                if (id != null && id.toLowerCase().contains("elser")) {
                    elserEndpoints.add(ep);
                }
            }

            if (!elserEndpoints.isEmpty()) {
                // Prefer endpoints starting with .elser-2- or .elser_model_2
                Optional<Map<String, Object>> preferred = elserEndpoints.stream()
                    .filter(ep -> {
                        String id = (String) ep.get("inference_id");
                        return id != null && (id.contains(".elser-2-") || id.contains(".elser_model_2"));
                    })
                    .findFirst();

                if (preferred.isPresent()) {
                    this.inferenceEndpoint = (String) preferred.get().get("inference_id");
                } else {
                    this.inferenceEndpoint = (String) elserEndpoints.get(0).get("inference_id");
                }

                logger.warning("Specified endpoint not found, using auto-detected: " + this.inferenceEndpoint);
                return true;
            }

            logger.severe("Inference endpoint '" + this.inferenceEndpoint + "' not found");
            logger.info("Available endpoints:");
            for (Map<String, Object> ep : endpoints) {
                String id = (String) ep.get("inference_id");
                if (id != null) {
                    logger.info("  - " + id);
                }
            }
            return false;
        } catch (Exception e) {
            logger.warning("Error checking inference endpoint: " + e.getMessage());
            logger.warning("Continuing anyway...");
            return true;
        }
    }

    public boolean createPipeline() {
        try {
            Map<String, Object> pipelineConfig = new HashMap<>();
            pipelineConfig.put("description", "Extract text from PDF - semantic_text field handles chunking and embeddings");

            List<Map<String, Object>> processors = new ArrayList<>();

            // Attachment processor
            Map<String, Object> attachmentProcessor = new HashMap<>();
            Map<String, Object> attachmentConfig = new HashMap<>();
            attachmentConfig.put("field", "data");
            attachmentConfig.put("target_field", "attachment");
            attachmentConfig.put("remove_binary", true);
            attachmentProcessor.put("attachment", attachmentConfig);
            processors.add(attachmentProcessor);

            // Set processor for semantic_content
            Map<String, Object> setProcessor1 = new HashMap<>();
            Map<String, Object> setConfig1 = new HashMap<>();
            setConfig1.put("field", "semantic_content");
            setConfig1.put("copy_from", "attachment.content");
            setConfig1.put("ignore_empty_value", true);
            setProcessor1.put("set", setConfig1);
            processors.add(setProcessor1);

            // Remove processor for data field
            Map<String, Object> removeProcessor = new HashMap<>();
            Map<String, Object> removeConfig = new HashMap<>();
            removeConfig.put("field", "data");
            removeConfig.put("ignore_missing", true);
            removeProcessor.put("remove", removeConfig);
            processors.add(removeProcessor);

            // Set processor for upload_date
            Map<String, Object> setProcessor2 = new HashMap<>();
            Map<String, Object> setConfig2 = new HashMap<>();
            setConfig2.put("field", "upload_date");
            setConfig2.put("value", "{{ _ingest.timestamp }}");
            setProcessor2.put("set", setConfig2);
            processors.add(setProcessor2);

            pipelineConfig.put("processors", processors);

            client.createPipeline(PIPELINE_NAME, pipelineConfig);
            return true;
        } catch (Exception e) {
            logger.severe("Error creating pipeline: " + e.getMessage());
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    public boolean createIndex() {
        try {
            // Delete index if it exists before creating a new one
            if (client.indexExists(ES_INDEX)) {
                logger.info("Deleting existing index '" + ES_INDEX + "' before import");
                if (client.deleteIndex(ES_INDEX)) {
                    logger.info("Index '" + ES_INDEX + "' deleted");
                } else {
                    logger.warning("Failed to delete index '" + ES_INDEX + "'");
                }
            }

            // Update mapping with detected inference endpoint
            Map<String, Object> mappingWithInference = deepCopy(mapping);
            if (mappingWithInference.containsKey("mappings")) {
                Map<String, Object> mappings = (Map<String, Object>) mappingWithInference.get("mappings");
                if (mappings != null && mappings.containsKey("properties")) {
                    Map<String, Object> properties = (Map<String, Object>) mappings.get("properties");
                    if (properties != null && properties.containsKey("semantic_content")) {
                        Map<String, Object> semanticContent = (Map<String, Object>) properties.get("semantic_content");
                        if (semanticContent != null) {
                            semanticContent.put("inference_id", this.inferenceEndpoint);
                        }
                    }
                }
            }

            logger.info("Creating index: " + ES_INDEX);
            client.createIndex(ES_INDEX, mappingWithInference);
            logger.info("Successfully created index: " + ES_INDEX);
            return true;
        } catch (Exception e) {
            logger.severe("Error creating index: " + e.getMessage());
            return false;
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> deepCopy(Map<String, Object> original) {
        ObjectMapper mapper = new ObjectMapper();
        try {
            String json = mapper.writeValueAsString(original);
            return mapper.readValue(json, Map.class);
        } catch (Exception e) {
            logger.warning("Failed to deep copy mapping, using original: " + e.getMessage());
            return new HashMap<>(original);
        }
    }

    private String extractAirlineName(String filename) {
        String filenameLower = filename.toLowerCase();

        if (filenameLower.contains("american")) {
            return "American Airlines";
        } else if (filenameLower.contains("southwest")) {
            return "Southwest";
        } else if (filenameLower.contains("united")) {
            return "United";
        } else if (filenameLower.contains("delta") || filenameLower.contains("dl-")) {
            return "Delta";
        } else {
            return "Unknown";
        }
    }

    private List<Path> getPdfFiles(String path) throws IOException {
        Path pathObj = Paths.get(path);

        if (!Files.exists(pathObj)) {
            logger.severe("Path '" + path + "' does not exist");
            return Collections.emptyList();
        }

        if (Files.isRegularFile(pathObj)) {
            String filename = pathObj.getFileName().toString().toLowerCase();
            if (filename.endsWith(".pdf")) {
                return Collections.singletonList(pathObj);
            } else {
                logger.severe("'" + path + "' is not a PDF file");
                return Collections.emptyList();
            }
        } else if (Files.isDirectory(pathObj)) {
            List<Path> pdfFiles = new ArrayList<>();
            try (Stream<Path> paths = Files.list(pathObj)) {
                paths.filter(Files::isRegularFile)
                    .filter(p -> p.getFileName().toString().toLowerCase().endsWith(".pdf"))
                    .sorted()
                    .forEach(pdfFiles::add);
            }
            if (pdfFiles.isEmpty()) {
                logger.warning("No PDF files found in directory '" + path + "'");
            }
            return pdfFiles;
        } else {
            return Collections.emptyList();
        }
    }

    private boolean indexPdf(Path pdfPath) {
        String filename = pdfPath.getFileName().toString();
        String airline = extractAirlineName(filename);

        try {
            // Read and encode the PDF
            byte[] pdfData = Files.readAllBytes(pdfPath);
            String encodedPdf = Base64.getEncoder().encodeToString(pdfData);

            // Index the document
            Map<String, Object> document = new HashMap<>();
            document.put("data", encodedPdf);
            document.put("filename", filename);
            document.put("airline", airline);

            client.indexDocument(ES_INDEX, document, PIPELINE_NAME);

            logger.info("Indexed: " + filename + " (airline: " + airline + ")");
            indexedCount++;
            return true;
        } catch (Exception e) {
            logger.severe("Error processing " + filename + ": " + e.getMessage());
            if (e.getCause() != null) {
                logger.severe("Cause: " + e.getCause().getMessage());
            }
            e.printStackTrace();
            return false;
        }
    }

    public boolean ingestPdfs(String pdfPath) {
        try {
            List<Path> pdfFiles = getPdfFiles(pdfPath);

            if (pdfFiles.isEmpty()) {
                logger.severe("No PDF files to process");
                return false;
            }

            logger.info("Processing " + pdfFiles.size() + " PDF file(s)...");

            int successCount = 0;
            int failedCount = 0;

            for (Path pdfFile : pdfFiles) {
                if (indexPdf(pdfFile)) {
                    successCount++;
                } else {
                    failedCount++;
                }
            }

            logger.info("Indexed " + successCount + " of " + pdfFiles.size() + " file(s)");
            if (failedCount > 0) {
                logger.warning("Failed: " + failedCount);
            }

            return failedCount == 0;
        } catch (Exception e) {
            logger.severe("Error ingesting PDFs: " + e.getMessage());
            return false;
        }
    }

    public boolean verifyIngestion() {
        try {
            // Add a small delay to ensure all documents are searchable
            Thread.sleep(1000);
            
            long count = client.countDocuments(ES_INDEX);
            logger.info("Index '" + ES_INDEX + "' contains " + count + " document(s)");
            
            if (count == 0 && indexedCount > 0) {
                logger.warning("Warning: Expected " + indexedCount + " document(s) but count shows 0. " +
                    "Documents may have failed during pipeline processing. Check Elasticsearch logs for details.");
            }
            
            return true;
        } catch (Exception e) {
            logger.warning("Could not verify document count: " + e.getMessage());
            return true;
        }
    }
}
